File: scripts/test-ocr-worker.js
/**
 * 测试 OCR Worker /extract-text 接口
 * 用法：node scripts/test-ocr-worker.js
 */

const SUPABASE_URL = process.env.NEXT_PUBLIC_SUPABASE_URL || process.env.SUPABASE_URL;
const SERVICE_ROLE_KEY = process.env.SUPABASE_SERVICE_ROLE_KEY;
const WORKER_BASE_URL = process.env.WORKER_BASE_URL || "https://pre-safe-scan.oook.cn";
const HMAC_SECRET = process.env.WORKER_HMAC_SECRET;

if (!SUPABASE_URL || !SERVICE_ROLE_KEY) {
  console.error("[ERROR] 缺少 SUPABASE_URL 或 SUPABASE_SERVICE_ROLE_KEY 环境变量");
  process.exit(1);
}
if (!HMAC_SECRET) {
  console.error("[ERROR] 缺少 WORKER_HMAC_SECRET 环境变量");
  process.exit(1);
}

// HMAC-SHA256 签名
async function hmacSign(secret, canonical) {
  const enc = new TextEncoder();
  const key = await crypto.subtle.importKey(
    "raw",
    enc.encode(secret),
    { name: "HMAC", hash: "SHA-256" },
    false,
    ["sign"]
  );
  const sig = await crypto.subtle.sign("HMAC", key, enc.encode(canonical));
  return Array.from(new Uint8Array(sig)).map(b => b.toString(16).padStart(2, "0")).join("");
}

// 生成 Supabase Storage 签名 URL（用 REST API，不引入 npm 包）
async function getSignedUrl(bucket, path, expiresIn = 3600) {
  const url = `${SUPABASE_URL}/storage/v1/object/sign/${bucket}/${path}`;
  const res = await fetch(url, {
    method: "POST",
    headers: {
      "Authorization": `Bearer ${SERVICE_ROLE_KEY}`,
      "Content-Type": "application/json",
      "apikey": SERVICE_ROLE_KEY,
    },
    body: JSON.stringify({ expiresIn }),
  });
  const data = await res.json();
  if (!res.ok || !data.signedURL) {
    throw new Error(`生成签名 URL 失败: ${JSON.stringify(data)}`);
  }
  return `${SUPABASE_URL}/storage/v1${data.signedURL}`;
}

async function testFile(label, storagePath, bucket = "dd-files") {
  console.log(`\n${"=".repeat(60)}`);
  console.log(`[TEST] ${label}`);
  console.log(`  storage: ${storagePath}`);

  // 1. 生成签名 URL
  let signedUrl;
  try {
    signedUrl = await getSignedUrl(bucket, storagePath);
    console.log(`  [OK] 签名 URL 生成成功`);
  } catch (err) {
    console.error(`  [FAIL] 签名 URL 生成失败:`, err.message);
    return;
  }

  // 2. 调用 Worker /extract-text
  const body = JSON.stringify({ file_url: signedUrl, max_chars: 3000 });
  const timestamp = Math.floor(Date.now() / 1000).toString();
  const nonce = crypto.randomUUID();
  const canonical = `${timestamp}.${nonce}.${body}`;
  const signature = await hmacSign(HMAC_SECRET, canonical);

  const start = Date.now();
  let res, data;
  try {
    res = await fetch(`${WORKER_BASE_URL.replace(/\/$/, "")}/extract-text`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "X-Timestamp": timestamp,
        "X-Nonce": nonce,
        "X-Signature": signature,
      },
      body,
    });
    data = await res.json();
  } catch (err) {
    console.error(`  [FAIL] 网络请求失败:`, err.message);
    return;
  }

  const elapsed = Date.now() - start;
  console.log(`  HTTP ${res.status} (${elapsed}ms)`);

  if (!res.ok || data.status === "error") {
    console.error(`  [FAIL] Worker 返回错误:`, JSON.stringify(data.error || data, null, 2));
    return;
  }

  const r = data.result;
  console.log(`  [OK] 提取成功`);
  console.log(`    extraction_source : ${r.extraction_source}`);
  console.log(`    file_type         : ${r.file_type}`);
  console.log(`    page_count        : ${r.page_count}`);
  console.log(`    ocr_fallback      : ${r.ocr_fallback_used}`);
  console.log(`    text_length       : ${(r.text || "").length} 字符`);
  console.log(`    text_preview      : ${(r.text || "").substring(0, 150).replace(/\n/g, " ")}...`);
}

async function main() {
  console.log(`Worker  : ${WORKER_BASE_URL}`);
  console.log(`Supabase: ${SUPABASE_URL}`);

  // 测试 PDF
  await testFile(
    "PDF",
    "e894040d-fbd7-4b3a-ab45-39abfc5525bd/2026-03-23/7pzunb4c.pdf"
  );

  // 测试图片 PNG
  await testFile(
    "PNG (image)",
    "e894040d-fbd7-4b3a-ab45-39abfc5525bd/2026-03-23/c8wrmwv0.png"
  );

  // 测试 DOCX
  await testFile(
    "DOCX",
    "e894040d-fbd7-4b3a-ab45-39abfc5525bd/2026-03-23/75m1ky3d.docx"
  );

  console.log(`\n${"=".repeat(60)}`);
  console.log("[DONE] 测试完成");
}

main().catch(err => {
  console.error("[FATAL]", err);
  process.exit(1);
});

File: scripts/test_ocr_worker.py
"""
测试 OCR Worker /extract-text 接口
"""
import hmac
import hashlib
import json
import os
import sys
import time
import uuid
import urllib.request
import urllib.error

SUPABASE_URL = (os.environ.get("NEXT_PUBLIC_SUPABASE_URL") or os.environ.get("SUPABASE_URL", "")).rstrip("/")
SERVICE_ROLE_KEY = os.environ.get("SUPABASE_SERVICE_ROLE_KEY", "")
WORKER_BASE_URL = (os.environ.get("WORKER_BASE_URL") or "https://pre-safe-scan.oook.cn").rstrip("/")
HMAC_SECRET = os.environ.get("WORKER_HMAC_SECRET", "")

print("=== 环境检查 ===")
print(f"WORKER_BASE_URL  : {WORKER_BASE_URL}")
print(f"SUPABASE_URL     : {SUPABASE_URL}")
print(f"HMAC_SECRET set  : {bool(HMAC_SECRET)}")
print(f"SERVICE_ROLE_KEY : {'set' if SERVICE_ROLE_KEY else 'MISSING'}")

if not HMAC_SECRET:
    print("[ERROR] 缺少 WORKER_HMAC_SECRET")
    sys.exit(1)

def hmac_sign(secret: str, canonical: str) -> str:
    return hmac.new(secret.encode(), canonical.encode(), hashlib.sha256).hexdigest()

def get_signed_url(bucket: str, path: str, expires_in: int = 3600) -> str:
    url = f"{SUPABASE_URL}/storage/v1/object/sign/{bucket}/{path}"
    payload = json.dumps({"expiresIn": expires_in}).encode()
    req = urllib.request.Request(
        url, data=payload, method="POST",
        headers={
            "Authorization": f"Bearer {SERVICE_ROLE_KEY}",
            "apikey": SERVICE_ROLE_KEY,
            "Content-Type": "application/json",
        }
    )
    try:
        with urllib.request.urlopen(req, timeout=15) as resp:
            data = json.loads(resp.read())
    except urllib.error.HTTPError as e:
        raise Exception(f"Supabase Storage HTTP {e.code}: {e.read().decode()[:200]}")
    
    signed = data.get("signedURL") or data.get("signedUrl")
    if not signed:
        raise Exception(f"未获取到 signedURL，响应: {json.dumps(data)[:200]}")
    # signedURL 已经包含完整路径（/storage/v1/object/sign/...?token=...）
    return f"{SUPABASE_URL}/storage/v1{signed}"

def call_worker(file_url: str, max_chars: int = 3000) -> dict:
    body = json.dumps({"file_url": file_url, "max_chars": max_chars})
    timestamp = str(int(time.time()))
    nonce = str(uuid.uuid4())
    canonical = f"{timestamp}.{nonce}.{body}"
    signature = hmac_sign(HMAC_SECRET, canonical)

    req = urllib.request.Request(
        f"{WORKER_BASE_URL}/extract-text",
        data=body.encode(),
        method="POST",
        headers={
            "Content-Type": "application/json",
            "X-Timestamp": timestamp,
            "X-Nonce": nonce,
            "X-Signature": signature,
        }
    )
    start = time.time()
    try:
        with urllib.request.urlopen(req, timeout=60) as resp:
            elapsed = int((time.time() - start) * 1000)
            data = json.loads(resp.read())
            return {"ok": True, "data": data, "status": resp.status, "elapsed": elapsed}
    except urllib.error.HTTPError as e:
        elapsed = int((time.time() - start) * 1000)
        err_body = e.read().decode()
        return {"ok": False, "status": e.code, "elapsed": elapsed, "error": err_body}
    except Exception as e:
        elapsed = int((time.time() - start) * 1000)
        return {"ok": False, "status": 0, "elapsed": elapsed, "error": str(e)}

def test_file(label: str, storage_path: str, bucket: str = "dd-files"):
    print(f"\n--- {label} ---")
    print(f"  storage_path: {storage_path}")

    try:
        signed_url = get_signed_url(bucket, storage_path)
        print(f"  [OK] 签名 URL: {signed_url[:80]}...")
    except Exception as e:
        print(f"  [FAIL] 签名 URL 错误: {e}")
        return

    print(f"  正在调用 Worker...")
    result = call_worker(signed_url)
    print(f"  HTTP {result['status']} ({result['elapsed']}ms)")

    if not result["ok"]:
        print(f"  [FAIL] {result.get('error', '')[:400]}")
        return

    data = result["data"]
    if data.get("status") == "error":
        err = data.get("error", data)
        print(f"  [FAIL] Worker error: {json.dumps(err, ensure_ascii=False)[:300]}")
        return

    r = data.get("result", {})
    text = r.get("text", "")
    print(f"  [OK] extraction_source : {r.get('extraction_source')}")
    print(f"       file_type         : {r.get('file_type')}")
    print(f"       page_count        : {r.get('page_count')}")
    print(f"       ocr_fallback      : {r.get('ocr_fallback_used')}")
    print(f"       text_length       : {len(text)} 字符")
    preview = text[:200].replace("\n", " ")
    print(f"       preview           : {preview}")

# 从 DB 查到的真实文件
test_file("PDF",  "e894040d-fbd7-4b3a-ab45-39abfc5525bd/2026-03-23/7pzunb4c.pdf")
test_file("PNG",  "e894040d-fbd7-4b3a-ab45-39abfc5525bd/2026-03-23/c8wrmwv0.png")
test_file("DOCX", "e894040d-fbd7-4b3a-ab45-39abfc5525bd/2026-03-23/75m1ky3d.docx")

print("\n=== 测试完成 ===")
